File: compiler/bench/factors.fs
\ expected: 343614
\ Factorize 100K numbers
\ Checksum: sum of number of prime factors for each

: count-factors ( n -- count )
  dup 2 < if drop 0 exit then
  0 swap
  2
  begin over 1 > while
    2dup mod 0= if
      swap over / swap
      rot 1+ -rot
    else
      dup 2 = if 1+ else 2 + then
    then
    2dup dup * < if
      over 1 > if rot 1+ -rot then
      1 swap
    then
  repeat 2drop ;

: bench-factors ( -- sum )
  0
  100000 0 do
    i 1+ count-factors +
  loop ;

: main bench-factors . cr ;

File: compiler/bench/factors.c
// expected: 343614
// Factorize 100K numbers
// Checksum: sum of number of prime factors for each

#include <stdio.h>

int count_factors(long n) {
    if (n < 2) return 0;
    int count = 0;
    for (long d = 2; d * d <= n; ) {
        if (n % d == 0) {
            count++;
            n /= d;
        } else {
            d += (d == 2) ? 1 : 2;
        }
    }
    if (n > 1) count++;
    return count;
}

int main(void) {
    long sum = 0;
    for (int i = 0; i < 100000; i++) {
        sum += count_factors(i + 1);
    }
    printf("%ld\n", sum);
    return 0;
}
